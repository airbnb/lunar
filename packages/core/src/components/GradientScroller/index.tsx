import React from 'react';
import { StyleBlock } from 'aesthetic';
import debounce from 'lodash/debounce';
import throttle from 'lodash/throttle';
import IconChevronLeft from '@airbnb/lunar-icons/lib/interface/IconChevronLeft';
import IconChevronRight from '@airbnb/lunar-icons/lib/interface/IconChevronRight';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import DirectionalIcon from '../DirectionalIcon';

const INCREMENT = 25;

export type Props = {
  /** How much to scroll when hovering above gradient. Defaults to 15% of the container width. */
  autoScrollIncrement?: number;
  /** Children to be positioned in a vertically centered fashion. */
  children?: React.ReactNode;
  /** Whether to hide the scrollbar. Doesn't prevent scroll. */
  hideScrollbar?: boolean;
  /** Whether to show arrows on the edges. */
  showArrows?: boolean;
};

export type State = {
  showStartGradient: boolean;
  showEndGradient: boolean;
};

/** A horizontal scroller with gradients on each side. */
export class GradientScroller extends React.Component<Props & WithStylesProps, State> {
  static defaultProps = {
    children: null,
    hideScrollbar: false,
    showArrows: false,
  };

  contentsRef: HTMLDivElement | null = null;

  observer: ResizeObserver;

  scrollInterval: number = 0;

  scrollerRef: HTMLDivElement | null = null;

  state = {
    showStartGradient: false,
    showEndGradient: true,
  };

  constructor(props: Props & WithStylesProps) {
    super(props);

    // Register resize observer before mounting
    this.observer = new window.ResizeObserver(this.handleObserver);
  }

  componentWillUnmount() {
    this.observer.disconnect();
  }

  calculate() {
    const scroller = this.scrollerRef;
    const contents = this.contentsRef;

    if (!scroller || !contents) {
      return;
    }

    const scrollerWidth = scroller.clientWidth;
    const maxChildWidth = this.calculateMaxChildWidth(contents);

    this.setState({
      showStartGradient: scroller.scrollLeft > 0,
      showEndGradient: maxChildWidth > scrollerWidth,
    });
  }

  calculateAutoScrollIncrement(): number {
    let increment = this.props.autoScrollIncrement;

    // 15% of the width
    if (!increment && this.scrollerRef) {
      increment = this.scrollerRef.offsetWidth * 0.15;
    }

    if (!increment) {
      increment = INCREMENT;
    }

    return increment;
  }

  calculateMaxChildWidth(contents: Element) {
    return Array.from(contents.children).reduce(
      (width, child) => Math.max(child.clientWidth, width),
      contents.clientWidth,
    );
  }

  doScroll(amount: number, interval: boolean = true) {
    const callback = () => {
      if (this.scrollerRef) {
        this.scrollerRef.scrollLeft += amount;
        this.handleScrollThrottled();
      }
    };

    this.handleScrollStop();

    if (interval) {
      this.scrollInterval = window.setInterval(callback, 100);
    } else {
      callback();
    }
  }

  private handleContentsRef = (ref: HTMLDivElement) => {
    if (ref) {
      this.observer.observe(ref);
    } else if (this.contentsRef) {
      this.observer.unobserve(this.contentsRef);
    }

    this.contentsRef = ref;
  };

  private handleScrollerRef = (ref: HTMLDivElement) => {
    if (ref) {
      this.observer.observe(ref);
    } else if (this.scrollerRef) {
      this.observer.unobserve(this.scrollerRef);
    }

    this.scrollerRef = ref;
  };

  private handleObserver = (entries: ResizeObserverEntry[]) => {
    let calculate = false;

    entries.forEach(({ contentRect, target }) => {
      const width = target.dataset.prevWidth;
      const nextWidth = String(contentRect.width);

      // Dataset stores as strings
      if (!width || nextWidth !== width) {
        calculate = true;

        // eslint-disable-next-line no-param-reassign
        target.dataset.prevWidth = nextWidth;
      }
    });

    if (calculate) {
      this.handleResizeDebounced();
    }
  };

  private handleResize = () => {
    this.calculate();
  };

  private handleResizeDebounced = debounce(this.handleResize, 150);

  private handleScroll = () => {
    const target = this.scrollerRef;
    const { showStartGradient, showEndGradient } = this.state;

    if (!target) {
      return;
    }

    if (target.scrollLeft > 0 && !showStartGradient) {
      this.setState({
        showStartGradient: true,
      });
    } else if (target.scrollLeft === 0 && showStartGradient) {
      this.setState({
        showStartGradient: false,
      });
    }

    // Get the largest width child to calculate against
    const scrolledWidth = target.scrollLeft + target.clientWidth;
    const maxChildWidth = this.calculateMaxChildWidth(target.children[0]);

    if (scrolledWidth < maxChildWidth && !showEndGradient) {
      this.setState({
        showEndGradient: true,
      });
    } else if (scrolledWidth >= maxChildWidth && showEndGradient) {
      this.setState({
        showEndGradient: false,
      });
    }
  };

  private handleScrollThrottled = throttle(this.handleScroll, 100);

  private handleScrollLeft = (event: React.MouseEvent) => {
    this.doScroll(-this.calculateAutoScrollIncrement(), event.type !== 'click');
  };

  private handleScrollRight = (event: React.MouseEvent) => {
    this.doScroll(this.calculateAutoScrollIncrement(), event.type !== 'click');
  };

  private handleScrollStop = () => {
    if (this.scrollInterval) {
      window.clearInterval(this.scrollInterval);
    }
  };

  render() {
    const { cx, children, styles, hideScrollbar, showArrows } = this.props;
    const { showStartGradient, showEndGradient } = this.state;

    return (
      <div className={cx(styles.container)}>
        <div className={cx(styles.leftGradient, showStartGradient && styles.gradient_reveal)}>
          {showArrows ? (
            <button
              className={cx(styles.leftArrow, hideScrollbar && styles.arrow_hideScrollbar)}
              type="button"
              onClick={this.handleScrollLeft}
            >
              <DirectionalIcon
                direction="left"
                left={IconChevronLeft}
                right={IconChevronRight}
                size="2em"
                decorative
              />
            </button>
          ) : (
            <span
              className={cx(styles.scrollTrigger)}
              onMouseEnter={this.handleScrollLeft}
              onMouseLeave={this.handleScrollStop}
            />
          )}
        </div>

        <div className={cx(styles.rightGradient, showEndGradient && styles.gradient_reveal)}>
          {showArrows ? (
            <button
              className={cx(styles.rightArrow, hideScrollbar && styles.arrow_hideScrollbar)}
              type="button"
              onClick={this.handleScrollRight}
            >
              <DirectionalIcon
                direction="right"
                left={IconChevronLeft}
                right={IconChevronRight}
                size="2em"
                decorative
              />
            </button>
          ) : (
            <span
              className={cx(styles.scrollTrigger)}
              onMouseEnter={this.handleScrollRight}
              onMouseLeave={this.handleScrollStop}
            />
          )}
        </div>

        <div
          className={cx(styles.scroller, hideScrollbar && styles.scroller_hideScrollbar)}
          ref={this.handleScrollerRef}
          onScroll={this.handleScrollThrottled}
        >
          <div className={cx(styles.contents)} ref={this.handleContentsRef}>
            {children}
          </div>
        </div>
      </div>
    );
  }
}

export default withStyles(({ color, unit, ui, pattern, transition }) => {
  const scrollbarHeight = unit * 1.5;

  const gradient: StyleBlock = {
    ...transition.fade,
    position: 'absolute',
    top: 0,
    bottom: 0,
    zIndex: 1,
    content: "''",
    opacity: 0,
    pointerEvents: 'none',
  };

  const arrow: StyleBlock = {
    ...pattern.resetButton,
    position: 'absolute',
    top: '45%',
    marginTop: -scrollbarHeight / 2,
    transform: 'translateY(-50%)',
    backgroundColor: color.accent.bg,
    borderRadius: '50%',
    boxShadow: ui.boxShadow,
    padding: unit,
    outline: 'none',

    ':hover': {
      backgroundColor: color.accent.bgHover,
    },
  };

  return {
    container: {
      width: '100%',
      position: 'relative',
      margin: -unit / 2,
    },

    leftGradient: {
      ...gradient,
      width: unit * 5,
      left: 0,
      background: `linear-gradient(to right, ${color.accent.bg} 40%, transparent 100%)`,
    },

    leftArrow: {
      ...arrow,
      left: -unit,
    },

    rightGradient: {
      ...gradient,
      width: unit * 5,
      right: 0,
      background: `linear-gradient(to left, ${color.accent.bg} 40%, transparent 100%)`,
    },

    rightArrow: {
      ...arrow,
      right: -unit,
    },

    arrow_hideScrollbar: {
      top: '45%',
      marginTop: 0,
    },

    gradient_reveal: {
      opacity: 1,
      pointerEvents: 'auto',
    },

    scroller: {
      overflowX: 'auto',
      overflowY: 'hidden',
      display: 'flex',
      alignItems: 'center',
      alignContent: 'stretch',
      paddingBottom: scrollbarHeight,
      '-webkit-overflow-scrolling': 'touch',
    },

    scroller_hideScrollbar: {
      paddingBottom: 0,

      '::-webkit-scrollbar': {
        display: 'none',
      },
    },

    scrollTrigger: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },

    contents: {
      flexBasis: '100%',
      padding: unit / 2,
    },
  };
})(GradientScroller);
