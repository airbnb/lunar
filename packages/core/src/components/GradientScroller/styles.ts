import { StyleBlock } from 'aesthetic';
import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, unit, ui, pattern, transition }) => {
  const scrollbarHeight = unit * 1.5;

  const gradient: StyleBlock = {
    ...transition.fade,
    position: 'absolute',
    top: 0,
    bottom: 0,
    zIndex: 1,
    content: "''",
    opacity: 0,
    pointerEvents: 'none',
  };

  const arrow: StyleBlock = {
    ...pattern.resetButton,
    position: 'absolute',
    top: '45%',
    marginTop: -scrollbarHeight / 2,
    transform: 'translateY(-50%)',
    backgroundColor: color.accent.bg,
    borderRadius: '50%',
    boxShadow: ui.boxShadow,
    padding: unit,
    outline: 'none',

    ':hover': {
      backgroundColor: color.accent.bgHover,
    },
  };

  return {
    container: {
      width: '100%',
      position: 'relative',
      margin: -unit / 2,
    },

    leftGradient: {
      ...gradient,
      width: unit * 5,
      left: 0,
      background: `linear-gradient(to right, ${color.accent.bg} 40%, transparent 100%)`,
    },

    leftArrow: {
      ...arrow,
      left: -unit,
    },

    rightGradient: {
      ...gradient,
      width: unit * 5,
      right: 0,
      background: `linear-gradient(to left, ${color.accent.bg} 40%, transparent 100%)`,
    },

    rightArrow: {
      ...arrow,
      right: -unit,
    },

    arrow_hideScrollbar: {
      top: '45%',
      marginTop: 0,
    },

    gradient_reveal: {
      opacity: 1,
      pointerEvents: 'auto',
    },

    scroller: {
      overflowX: 'auto',
      overflowY: 'hidden',
      display: 'flex',
      alignItems: 'center',
      alignContent: 'stretch',
      paddingBottom: scrollbarHeight,
      '-webkit-overflow-scrolling': 'touch',
    },

    scroller_hideScrollbar: {
      paddingBottom: 0,

      '::-webkit-scrollbar': {
        display: 'none',
      },
    },

    scrollTrigger: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },

    contents: {
      flexBasis: '100%',
      padding: unit / 2,
    },
  };
};

export default styleSheet;

export { styleSheet };
