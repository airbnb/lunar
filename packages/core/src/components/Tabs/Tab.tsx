import React from 'react';
import camelCase from 'lodash/camelCase';
import upperFirst from 'lodash/upperFirst';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import ButtonOrLink, { Props as ButtonOrLinkProps } from '../private/ButtonOrLink';
import TrackingBoundary from '../TrackingBoundary';

export type Props = Pick<ButtonOrLinkProps, 'afterIcon' | 'beforeIcon' | 'disabled' | 'href'> & {
  /** Hide bottom border of Tab when unselected. */
  borderless?: boolean;
  /** Content to render if the tab is selected. */
  children?: React.ReactNode;
  /**
   * Unique key name for this tab.
   * @ignore
   */
  keyName?: string;
  /** Text to render in the tab. */
  label: NonNullable<React.ReactNode>;
  /** Callback fired when the tab is clicked. */
  onClick?: (key: string) => void;
  /** Callback fired when the tab is selected. */
  onSelected?: () => void;
  /** Whether the tab is selected or not. */
  selected?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
  /** Stretch tabs to fill the full width. */
  stretched?: boolean;
};

/** A single tab button. Usually rendered amongst a collection of tabs. */
export class Tab extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    borderless: false,
    children: null,
    selected: false,
    small: false,
    stretched: false,
  };

  componentDidUpdate(prevProps: Props) {
    if (!prevProps.selected && this.props.selected && this.props.onSelected) {
      this.props.onSelected();
    }
  }

  private handleClick = () => {
    if (this.props.onClick) {
      this.props.onClick(this.props.keyName!);
    }
  };

  render() {
    const {
      cx,
      afterIcon,
      beforeIcon,
      borderless,
      disabled,
      href,
      keyName,
      label,
      selected,
      small,
      stretched,
      styles,
    } = this.props;
    const trackingName = upperFirst(camelCase(keyName || 'Tab'));

    return (
      <span
        className={cx(
          styles.tab,
          disabled && styles.tab_disabled,
          borderless && styles.tab_borderless,
          selected && styles.tab_selected,
          stretched && styles.tab_stretched,
        )}
      >
        <TrackingBoundary name={trackingName}>
          <ButtonOrLink
            flexAlign
            aria-selected={selected}
            afterIcon={afterIcon}
            beforeIcon={beforeIcon}
            disabled={disabled}
            href={href}
            role="tab"
            onClick={disabled ? undefined : this.handleClick}
            className={cx(
              styles.tabButton,
              small && styles.tabButton_small,
              selected && styles.tabButton_active,
            )}
          >
            {label}
          </ButtonOrLink>
        </TrackingBoundary>
      </span>
    );
  }
}

export default withStyles(({ color, font, pattern, unit, ui, transition }) => ({
  tab: {
    borderBottom: ui.borderThick,
    marginRight: unit * 4,
    marginBottom: -2,

    ':hover': {
      borderColor: color.accent.borderHover,
    },

    ':last-child': {
      marginRight: 0,
    },
  },

  tab_borderless: {
    borderColor: color.clear,
  },

  tab_selected: {
    borderColor: color.accent.borderActive,

    ':hover': {
      borderColor: color.accent.borderActive,
    },
  },

  tab_stretched: {
    flex: 1,
  },

  tab_disabled: {
    ...pattern.disabled,
    borderColor: color.accent.border,

    ':hover': {
      borderColor: color.accent.border,
    },
  },

  tabButton: {
    ...font.textReset,
    ...font.textRegular,
    ...transition.box,
    background: color.accent.bg,
    color: color.accent.text,
    display: 'flex',
    alignItems: 'center',
    paddingTop: unit,
    paddingBottom: unit,
    border: 0,
    cursor: 'pointer',
    textAlign: 'left',
    textDecoration: 'none',
    whiteSpace: 'nowrap',
    width: '100%',
    fontWeight: font.weights.bold,
    borderTopLeftRadius: ui.borderRadius,
    borderTopRightRadius: ui.borderRadius,
  },

  tabButton_small: {
    ...font.textSmall,
  },

  tabButton_active: {
    color: color.accent.textActive,
  },
}))(Tab);
