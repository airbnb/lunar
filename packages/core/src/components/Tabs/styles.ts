import { StyleSheet } from '../../hooks/useStyles';

export const styleSheet: StyleSheet = ({ color, ui, unit }) => ({
  nav: {
    borderBottom: ui.borderThick,
    display: 'flex',
  },

  nav_noBorder: {
    borderColor: color.clear,
  },

  nav_secondary: {
    alignItems: 'center',
    borderWidth: 0,
  },

  panel: {
    marginTop: unit,
  },
});

export const styleSheetTab: StyleSheet = ({ color, font, pattern, unit, ui, transition }) => ({
  tab: {
    borderBottom: ui.borderThick,
    marginRight: unit * 4,
    marginBottom: -2,

    ':hover': {
      borderColor: color.accent.borderHover,
    },

    ':last-child': {
      marginRight: 0,
    },
  },

  tab_secondary: {
    marginRight: unit,
    borderWidth: 0,
    marginBottom: 0,
  },

  tab_noBorder: {
    borderColor: color.clear,
  },

  tab_selected: {
    borderColor: color.accent.borderActive,

    ':hover': {
      borderColor: color.accent.borderActive,
    },
  },

  tab_stretched: {
    flex: 1,
  },

  tab_disabled: {
    borderColor: color.accent.border,

    ':hover': {
      borderColor: color.accent.border,
    },
  },

  tab_noHover: {
    ':hover': {
      borderColor: color.clear,
    },
  },

  tabButton: {
    ...font.textReset,
    ...font.textRegular,
    ...transition.box,
    background: color.clear,
    color: color.accent.text,
    display: 'flex',
    alignItems: 'center',
    paddingTop: unit,
    paddingBottom: unit,
    border: 0,
    cursor: 'pointer',
    textAlign: 'left',
    textDecoration: 'none',
    whiteSpace: 'nowrap',
    width: '100%',
    fontWeight: font.weights.bold,
    borderTopLeftRadius: ui.borderRadius,
    borderTopRightRadius: ui.borderRadius,
  },

  tabButton_secondary: {
    ...pattern.regularButton,
    paddingTop: unit / 2,
    paddingBottom: (unit / 8) * 5,
    fontWeight: 'normal',
    justifyContent: 'center',
    border: `${ui.borderWidth}px solid ${color.clear}`,
    backgroundColor: color.clear,
    borderRadius: ui.borderRadius,
    ':hover': {
      borderColor: color.accent.borderHover,
      backgroundColor: color.core.neutral[1],
    },
  },

  tabButton_selected: {
    color: color.accent.textActive,
  },

  tabButton_secondary_selected: {
    borderColor: color.accent.borderActive,
    backgroundColor: color.accent.bg,
    ':hover': {
      borderColor: color.core.primary[4],
      backgroundColor: color.accent.bgHover,
    },
  },

  tabButton_disabled: {
    ...pattern.disabled,
    pointerEvents: 'none',
  },

  tabButton_small: {
    ...font.textSmall,
  },
});
