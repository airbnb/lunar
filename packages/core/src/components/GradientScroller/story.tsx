import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Text from '../Text';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';
import GradientScroller from '.';

class ButtonGroupDemo extends React.Component<{}, { buttons: number[] }> {
  state = { buttons: [] };

  handleClick = () => {
    this.setState(prevState => ({
      buttons: [...prevState.buttons, prevState.buttons.length],
    }));
  };

  render() {
    return (
      <div style={{ borderBottom: '2px solid #DBDBDB', paddingBottom: 4 }}>
        <ButtonGroup>
          <Button onClick={this.handleClick}>Add</Button>

          {this.state.buttons.map(count => (
            <Button key={count}>{count}</Button>
          ))}
        </ButtonGroup>
      </div>
    );
  }
}

export default {
  title: 'Core/GradientScroller',
  parameters: {
    inspectComponents: [GradientScroller],
  },
};

export function aScrollableContainer() {
  return (
    <GradientScroller>
      <Text>
        <span style={{ whiteSpace: 'nowrap' }}>
          <LoremIpsum />
          <LoremIpsum />
        </span>
      </Text>
    </GradientScroller>
  );
}

aScrollableContainer.story = {
  name: 'A scrollable container.',
};

export function withLeftAndRightArrows() {
  return (
    <GradientScroller showArrows hideScrollbar>
      <Text>
        <span style={{ whiteSpace: 'nowrap' }}>
          <LoremIpsum />
          <LoremIpsum />
        </span>
      </Text>
    </GradientScroller>
  );
}

withLeftAndRightArrows.story = {
  name: 'With left and right arrows.',
};

export function withNoScrollbarAndVariableWidthChildren() {
  return (
    <GradientScroller hideScrollbar>
      <ButtonGroupDemo />
    </GradientScroller>
  );
}

withNoScrollbarAndVariableWidthChildren.story = {
  name: 'With no scrollbar and variable width children.',
  parameters: { happo: false },
};
