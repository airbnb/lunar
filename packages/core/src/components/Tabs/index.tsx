/* eslint-disable jsx-a11y/no-noninteractive-element-to-interactive-role */

import React from 'react';
import { childrenOfType } from 'airbnb-prop-types';
import withBoundary from '../../composers/withBoundary';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';
import GradientScroller from '../GradientScroller';
import Tab, { Props as TabProps } from './Tab';

export type Props = {
  /** Hide bottom border of Tabs. */
  borderless?: boolean;
  /** Tabs and their content. */
  children: NonNullable<React.ReactNode>;
  /** Key of tab selected by default. */
  defaultKey?: string;
  /** Callback fired when a tab changes. */
  onChange?: (key: string) => void;
  /** Persist the selected tab through the defined URL hash. */
  persistWithHash?: string;
  /** Wrap tabs in a scrollable region. */
  scrollable?: boolean;
  /** Stretch tabs to fill the full width. */
  stretched?: boolean;
  /** A unique name for tracking purposes. */
  trackingName?: string;
};

export type State = {
  selectedKey: string;
};

/** A controller for multiple tabs. */
export class Tabs extends React.Component<Props & WithStylesProps, State> {
  static propTypes = {
    children: childrenOfType(Tab).isRequired,
  };

  static defaultProps = {
    borderless: false,
    defaultKey: '',
    onChange() {},
    persistWithHash: '',
    scrollable: false,
    stretched: false,
  };

  state = {
    selectedKey: this.getDefaultSelectedKey(),
  };

  componentDidUpdate(prevProps: Props) {
    if (this.props.defaultKey !== prevProps.defaultKey) {
      this.setState({
        selectedKey: this.props.defaultKey!,
      });
    }
  }

  getDefaultSelectedKey(): string {
    const { defaultKey, persistWithHash } = this.props;

    return (persistWithHash && this.getHashQuery().get(persistWithHash)) || defaultKey || '';
  }

  getHashQuery(): URLSearchParams {
    const { hash } = location;

    return new URLSearchParams(hash.length > 1 ? hash.slice(1) : '');
  }

  private handleClick = (key: string) => {
    const { persistWithHash } = this.props;

    this.setState({
      selectedKey: key,
    });

    this.props.onChange!(key);

    if (persistWithHash) {
      const query = this.getHashQuery();

      query.set(persistWithHash, key);

      history.pushState(null, '', `#${String(query)}`);
    }
  };

  render() {
    const { borderless, children, scrollable, stretched, styles } = this.props;
    const { selectedKey } = this.state;

    // Generate content
    let content = null;
    const nav = (
      <nav role="tablist" {...css(styles.nav, borderless && styles.nav_borderless)}>
        {React.Children.map(children, (child, i) => {
          if (!child) {
            return null;
          }

          const { key, props } = child as any;
          const selected = key === selectedKey || (!selectedKey && i === 0);

          if (__DEV__ && !key) {
            throw new Error('Tab components require a unique `key`.');
          }

          if (selected && props && props.children) {
            content = props.children;
          }

          return React.cloneElement(child as React.ReactElement<TabProps>, {
            borderless,
            keyName: key,
            selected,
            stretched,
            onClick: this.handleClick,
          });
        })}
      </nav>
    );

    return (
      <div>
        {scrollable ? (
          <GradientScroller hideScrollbar showArrows>
            {nav}
          </GradientScroller>
        ) : (
          nav
        )}

        {content && (
          <section role="tabpanel" {...css(styles.panel)}>
            {content}
          </section>
        )}
      </div>
    );
  }
}

export { Tab };

export default withBoundary('Tabs')(
  withStyles(({ color, ui, unit }) => ({
    nav: {
      borderBottom: ui.borderThick,
      display: 'flex',
    },

    nav_borderless: {
      borderColor: color.clear,
    },

    panel: {
      marginTop: unit,
    },
  }))(Tabs),
);
