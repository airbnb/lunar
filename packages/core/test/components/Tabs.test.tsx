import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import { unwrapHOCs } from '@airbnb/lunar-test-utils';
import Tabs from '../../src/components/Tabs';
import Tab from '../../src/components/Tabs/Tab';
import GradientScroller from '../../src/components/GradientScroller';

function unwrap(element: any): Enzyme.ShallowWrapper {
  // Dont use shallowWithStyles because of TrackingBoundary
  return unwrapHOCs(shallow(element), 'Tabs', {}, { render: true });
}

describe('<Tabs/>', () => {
  it('errors if a tab does not have a `key`', () => {
    expect(() =>
      unwrap(
        <Tabs>
          <Tab label="One" />
        </Tabs>,
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('sets selected key state using `defaultKey`', () => {
    const wrapper = unwrap(
      <Tabs defaultKey="c">
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.state('selectedKey')).toBe('c');
  });

  it('sets selected key state using hash', () => {
    location.hash = '#tab=c';

    const wrapper = unwrap(
      <Tabs persistWithHash="tab">
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.state('selectedKey')).toBe('c');
  });

  it('doesnt set selected key state using hash if names dont match', () => {
    location.hash = '#othertab=c';

    const wrapper = unwrap(
      <Tabs persistWithHash="tab">
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.state('selectedKey')).toBe('');
  });

  it('passes selected key state to each tab', () => {
    const wrapper = unwrap(
      <Tabs defaultKey="b">
        <Tab key="a" label="One" />
        <Tab key="b" label="Two" />
        <Tab key="c" label="Three" />
      </Tabs>,
    );

    expect(
      wrapper
        .find(Tab)
        .at(0)
        .prop('selected'),
    ).toBe(false);
    expect(
      wrapper
        .find(Tab)
        .at(1)
        .prop('selected'),
    ).toBe(true);
    expect(
      wrapper
        .find(Tab)
        .at(2)
        .prop('selected'),
    ).toBe(false);
  });

  it('adds `keyName`s to tabs', () => {
    const wrapper = unwrap(
      <Tabs>
        <Tab key="a" label="One" />
        <Tab key="b" label="Two" />
        <Tab key="c" label="Three" />
      </Tabs>,
    );

    expect(
      wrapper
        .find(Tab)
        .at(0)
        .prop('keyName'),
    ).toBe('a');
    expect(
      wrapper
        .find(Tab)
        .at(1)
        .prop('keyName'),
    ).toBe('b');
    expect(
      wrapper
        .find(Tab)
        .at(2)
        .prop('keyName'),
    ).toBe('c');
  });

  it('it renders a nav', () => {
    const wrapper = unwrap(
      <Tabs>
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.find('nav')).toHaveLength(1);
  });

  it('it renders a stretched nav', () => {
    const wrapper = unwrap(
      <Tabs stretched>
        <Tab key="a" label="One" />
        <Tab key="b" label="Two" />
        <Tab key="c" label="Three" />
      </Tabs>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('it renders a section when a tab is active and has children', () => {
    const wrapper = unwrap(
      <Tabs>
        <Tab key="a" label="Label">
          Foo
        </Tab>
      </Tabs>,
    );

    expect(wrapper.find('section')).toHaveLength(1);
    expect(wrapper.find('section').text()).toBe('Foo');
  });

  it('doesnt render a section if the tab has no children', () => {
    const wrapper = unwrap(
      <Tabs>
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.find('section')).toHaveLength(0);
  });

  it('handles falsey tabs', () => {
    const wrapper = unwrap(
      <Tabs>
        {false && <Tab key="a" label="One" />}
        {null && <Tab key="b" label="Two" />}
        <Tab key="c" label="Three" />
      </Tabs>,
    );

    expect(wrapper.find(Tab)).toHaveLength(1);
  });

  it('updates state index when `defaultKey` changes', () => {
    const wrapper = unwrap(
      <Tabs defaultKey="c">
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.state('selectedKey')).toBe('c');

    wrapper.setProps({
      defaultKey: 'b',
    });

    expect(wrapper.state('selectedKey')).toBe('b');
  });

  it('triggers `onChange` when clicking', () => {
    const spy = jest.fn();
    const wrapper = unwrap(
      <Tabs onChange={spy}>
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    wrapper.find(Tab).simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  it('wraps in scroller when using `scrollable`', () => {
    const wrapper = unwrap(
      <Tabs scrollable>
        <Tab key="a" label="Label" />
      </Tabs>,
    );

    expect(wrapper.find(GradientScroller)).toHaveLength(1);
  });

  it('updated hash when tab is clicked', () => {
    location.hash = '#tab=a';

    const wrapper = unwrap(
      <Tabs persistWithHash="tab">
        <Tab key="a" label="Label" />
        <Tab key="b" label="Label" />
        <Tab key="c" label="Label" />
      </Tabs>,
    );

    expect(wrapper.state('selectedKey')).toBe('a');

    wrapper
      .find(Tab)
      .at(2)
      .simulate('click', 'c');

    expect(wrapper.state('selectedKey')).toBe('c');
    expect(location.hash).toBe('#tab=c');
  });

  it('it passes the borderless prop to Tab children', () => {
    const wrapper = unwrap(
      <Tabs borderless>
        <Tab key="a" label="One" />
      </Tabs>,
    );

    expect(
      wrapper
        .find(Tab)
        .at(0)
        .prop('borderless'),
    ).toBe(true);
  });
});
