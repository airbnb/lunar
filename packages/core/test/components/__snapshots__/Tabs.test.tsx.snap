// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs/> errors if a tab does not have a \`key\` 1`] = `"Tab components require a unique \`key\`."`;

exports[`<Tabs/> errors if non-tab children are passed 1`] = `
"Warning: Failed prop type: \`Tabs\` only accepts children of type withStyles(Tab)
    in Tabs"
`;

exports[`<Tabs/> it renders a stretched nav 1`] = `
<div>
  <nav
    className="nav_1x0pvrr"
    role="tablist"
  >
    <Tab
      borderless={false}
      key=".$a"
      keyName="a"
      label="One"
      onClick={[Function]}
      selected={true}
      stretched={true}
    />
    <Tab
      borderless={false}
      key=".$b"
      keyName="b"
      label="Two"
      onClick={[Function]}
      selected={false}
      stretched={true}
    />
    <Tab
      borderless={false}
      key=".$c"
      keyName="c"
      label="Three"
      onClick={[Function]}
      selected={false}
      stretched={true}
    />
  </nav>
</div>
`;
